### DISCUSSION TITLE
Overpowering with several heater

### DISCUSSION THREAD

**Discussion by adi90x** [jmcollin78/versatile_thermostat#178] [Ideas]
Hello,

Opening this to discuss about #174  and #150
With patch #174, overpowering may skipped if we have several heater of different power ( like one small and one big )  as it may trigger big heater while available power is only ```total_power/number of heater ``` ( i.e : average power), this may lead to over powering and potential issue. 
Do we want a solution or just update documentation to make user aware ? 
Possible solution :

1. Check for full power -> may trigger overpowering state while part of heater may be available
2. Improve this part by setting and using power for each underlying. 
3. Keep it as it is, add warning into documentation

Any other idea or comment on this ?

**Comment by jmcollin78** [comment]
Hello @adi90x ,

Check overpowering is not done at underlying level but at VTherm level. That why without breaking all things it is not so easy to have a perfect feature.
A perfect solution have never be the goal of this feature because heater power consumption is an maximum approximation given by the user. With my hvac, the power can go from 250 in cool mode and 1000 w in heat mode. The goal is to limitate the power consumption by shuting down heater when the limit is near to be reach.

It is obvious (for me) that the user must take some security margin. But you are right, may be it is just obvious in my head and the documentation should certainly be improved.

So my proposal is to improve the documentation with some warnings.

**Reply by adi90x** [reply]
Hello @jmcollin78 , 

Thanks a lot for your feedback, my issue is that with #174 we are not on the safe side.
It seems safer to not be heating than risking overpowering, no ? 
One other solution, could be to have sum of power as right now + max power of one heater ?

In fact, re-reading #174, I don't understand why we even check ```self._device_power * self._prop_algorithm.on_percent``` => In term of overpowering check for an eletric heater it make no sense, as heater is either on or off so full power needed or 0 power.
We will never reach that amount at any single point in time ( even if it will be the correct amount for power consumption of the device over the cycle )

If we do call this check before firing up each individual heater, then we could expect power to be ```self._device_power / self.nb_underlying_entities``` ( in the actual setup ) 

If we call it only once per cycle, then only power that make sense is full power no ? 

May need to re-read the code, but from what I recall if on_percent = 60% with 2 heater,  each heater will be on 60% of the time ( so at least 20% overlap where full power will be needed )  

Regards,

**Comment by jmcollin78** [comment]
> self._device_power * self._prop_algorithm.on_percent 

It is written in comment because I was sure that it will leads to questions...
If on_percent is 100%, the 4 heaters are on. So the power consumed is the total of `device_power`.

if it is not the case, you could have 1 or 2 or 3 or ... heaters running at the same time. This occurs when on_percent > 25% with 4 heaters. The best approximation, I found which is not a `usine à gas` of how much power is consumed during the cycle is `self._device_power * self._prop_algorithm.on_percent`. The value is correct during a cycle but can be overriden on a short period (during overlap of heating)

> as heater is either on or off so full power needed or 0 power

Yes. That is why i take the max between one full heater power and the value explained above. The value taken into account cannot be less than a full heater power (in mean value).

You talk about safety but THIS IS NOT a safety function. Don't put the safety of your home into this feature.

This is more to avoid too much consumption (pour faire des économies et lisser la charge, je ne sais pas comment dire ça en Anglais).

So a warning message in README if it is not already present is my prefered solution.

**Reply by adi90x** [reply]
Je passe aussi en francais, cela sera plus simple.
Mon point est que ``` self._device_power * self._prop_algorithm.on_percent ``` correspond parfaitement a l'idée de consommation moyenne, en revanche il ne va absolument pas servir pour du délestage/ ```eviter de faire sauter les plombs``` :)
En fait, je conçois que l'on utilise cela pour calculer une conso d'électricité mais dans le cas du overpowering il me semble plus logique de checker la consommation instantanée qui sera necessaire à l'allumage pour décider ou non de l'allumage et non la consommation du cycle, non ?

**Reply by adi90x** [reply]
Je me réponds à moi même ( sans savoir si cela résout mon probleme ou en crée un autre ) car lorsque le premier chauffage va démarrer, la conso total de la maison va changer ( augmenter de device_power/n ), donc un event va être fired et le overpowering check sera à nouveau effectué et tentera à nouveau d'ajouter device_power/n

EDIT :  ou alors je suis fatigué ?

**Reply by jmcollin78** [reply]
J'ai bien compris. Mais le pic de conso sur un cycle d'un VTherm à 4 radiateurs dont on n'a pas la puissance individuelle et avec un on_percent à 34 % il est de combien ? C'est ça qui est compliqué et que je refuse de calculer. Je veux faire des trucs simples et qui marchent. Pas des trucs parfaits sur des bases floues qui ne marchent pas ou que les gens ne comprennent pas.

Ce qui est facile à comprendre c'est qu'un radiateur de 1000 w allumé à 60% il va consommer en gros 600 w. 

Sachant en plus que la consommation de la maison ne remonte pas instantanée non plus. Y a forcément une latence donc le calcul il est flou et sur des valeurs fausses.

Le seul truc parfait, c'est l'équipement électronique branché sur le compteur qui fait du délestage en temps réel. Ca oui c'est de la sécurité. Je ne peux et ne veux pas faire ça à ce niveau.

Et je ne pense pas que couper le chauffage parce-que pendant 1 min on pourrait éventuellement se rapprocher de la marge soit une bonne chose non plus.

C'est une problème d'objectif: cette fonction elle sert à ne pas trop dépasser un seuil. Pas à sécuriser un logement en évitant de mettre le feu. Elle me sert typiquement, quand je recharge la voiture a moins faire tourner le chauffage (à tour de role les VTherm vont être délestés). Je ne veux pas ne plus avoir de chauffage quand je recharge ma voiture. Tu vois la différence ?

Dans mon cas, j'ai un abonnement 18kva et le seuil je l'ai mis à 12. Des fois, très rarement mais c'est possible, ca dépasse les 12. Parce-que y a pas que des VTherms dans ma maison. 

Donc un bon warning dans le README est une excellente proposition, car c'est pas forcément clair pour tout le monde.

**Reply by jmcollin78** [reply]
> EDIT : ou alors je suis fatigué ?

Il commence à se faire tard oui. Le seul moyen c'est avant d'allumer, vérifier si on a la puissance nécessaire: puissance totale encours + celle du radiateur que je veux allumer < seuil ?

**Reply by adi90x** [reply]
Je comprends totalement l'idée et je te rejoins sur la latence et la simplicité (et en effet prendre une bonne marge entre l'abonnement et le seuil semble nécessaire ).
Apres je reste sur mon idée que j'aurais préféré checker le pire cas possible ( donc le device_power ) plutôt qu'un estimé qui ne sera "jamais" la consommation "quasi-live" de mon vtherm :)

**Reply by adi90x** [reply]
> > EDIT : ou alors je suis fatigué ?
> 
> Il commence à se faire tard oui. Le seul moyen c'est avant d'allumer, vérifier si on a la puissance nécessaire: puissance totale encours + celle du radiateur que je veux allumer < seuil ?

Ce check est fait par cycle du vtherm ( lors de l'appel à ```async_control_heating``` ) , correct ? 

Mais lorsque l'on recoit un event de changement de puissance ( donc typiquement quelques secondes après que le premier radiateur a commencé son cycle ) : la fonction ```async_control_heating``` et donc check d'overpowering va à nouveau avoir lieu avec : ```puissance totale ( incluant maintenant 1 radiateur ) + device_power/n < seuil```  ? 

Si oui, alors cela répond à ma problématique tant que le temps entre déclechement de chaque radiateur au sein d'un cycle < latence remonté conso elec et je n'ai pas fait le calcul mais lorsque le temps de déclenchement entre les radiateurs se rapproche de la latence remonté de conso ( +-7sec ) le on_percent sera >> 25% ( dans le cas d'un scénario à 4 radiateurs ))

**Reply by jmcollin78** [reply]
Le check est fait lorsqu'on doit allumer un radiateur qui était éteint. C'est seulement à ce moment là qu'on a besoin d'un peu de puissance dispo. Donc comme il n'est pas encore allumé, sa conso n'est pas dans le capteur de puissance totale.

**Comment by adi90x** [comment]
Mais le check est bien fait à chaque fois que la conso live change non ?
Via l'event power change et l'appel au check de cycle non ?


Cordialement,

Le jeu. 9 nov. 2023, 08:59, Jean-Marc Collin ***@***.***> a
écrit :

**Comment by jmcollin78** [comment]
> Mais le check est bien fait à chaque fois que la conso live change non ?

Non pas tout le temps sinon on le ferait en permanence puisque la conso live change tout le temps. Le check est fait à chaque cycle ou lorsqu'on doit allumer / reallumer un radiateur (fin de window par exemple).

La seule exception, c'est quand le radiateur est déjà en mode overpowering où en effet on fait le check à chaque fois que la conso live change ou que la puissance max change pour remettre au plus tot le chauffage.
